### DISCUSSION TITLE
WebRTC/getUserMedia() returning undefined

### DISCUSSION THREAD

**Discussion by tohmeeey** [SafeExamBrowser/seb-mac#300] [Q&A]
I'm currently running the latest release for mac . My web application is trying to get permission for camera and microphone using the WebRTC/getUserMedia() API. But a typeerror is returmed amd navigator.mediaDevices is undefined. 

Note: The webapp works well outside SEB and navigator.mediaDevices is not undefined when using browsers outside SEB.
As per the documentation, I also have all the new settings browserMediaCaptureCamera and browserMediaCaptureMicrophone (Settings / Browser Features / Media Playback/Capture / Allow Camera/Microphone Capture) set to true.

Please advise on what to do.

**Comment by danschlet** [comment]
I would assume your web application isn't using the modern WebView in SEB. Make sure to follow instructions on [this page](https://safeexambrowser.org/developer/seb-config-key.html), section "Query the Config Key and Browser Exam Key with the SEB JavaScript API". You can test if SEB is using the modern WebView by trying if pinch-to-zoom with two fingers on a trackpad works (unless you deactivated page zoom using the according SEB setting).

**Reply by tohmeeey** [reply]
Thanks a lot for your response, it helped. I now have access to the microphone and camera. 

There's a follow-up problem now though. My web app has a feature that allows users to take pictures and upload, and although I have access to the camera and they can see themselves through the thumbnail, when they capture it comes out black. 
This is not happening when I use firefox.
Is there a setting affecting this?

**Reply by danschlet** [reply]
No, I don't think any SEB settings should affect this. But I know that I observed a similar issue in the Moodle rich text editor which also has a feature to take pictures, the preview works but there is no resulting picture (or video) taken. This also happens with a regular Safari browser, so I assume this may be a WebKit incompatibility of the involved web APIs. WebRTC based camera and microphone access nevertheless works without issues (for example tested with the BBB, Jitsi Meet and Zoom web clients).
Maybe you need to use another version of the getUserMedia API, as mentioned [here](https://stackoverflow.com/questions/63672276/navigator-getusermedia-not-working-on-safari-but-working-on-other-browsers)?

**Reply by tohmeeey** [reply]
Hi, my code is actually the same as what was recommended in the link you shared. So, I have an up-to-date implementation of webrtc. Is there anything else you could recommend? cos I just discovered the same thing is happening with recorded audio too. the mic picks up sounds but no recorded audio output is produced. And this is not happening outside of SEB
